### ISSUE TITLE
DnsNameResolver seems overly strict when validating service config

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc-java#6579]
DnsNameResolver only allows a pre-determined set of keys used as filters, _in the entire list_:
https://github.com/grpc/grpc-java/blob/c606519a5d75b844bfc8bd243c6c9ebe1c742f96/core/src/main/java/io/grpc/internal/DnsNameResolver.java#L543-L545

That means we can never add a new filter key. That seems like a very bad idea. I don't know if this is cross-language, but the precise behavior doesn't seem nailed down in https://github.com/grpc/proposal/blob/master/A2-service-configs-in-dns.md .

CC @dfawley, @markdroth

**Comment by markdroth** [comment]
I agree that this is very wrong.  The decision to ignore unknown fields was specified in the service config error handling design (https://github.com/grpc/proposal/blob/master/A21-service-config-error-handling.md#criteria-to-determine-the-validity-of-a-service-config).

C-core does properly allow unknown fields (https://github.com/grpc/grpc/blob/4f0875dd9e38006d89d4a0cd122280472f58f7c5/src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc#L235).

**Comment by ejona86** [comment]
@markdroth, as a language lawyer I'll note that the DNS selection filters stuff isn't technically part of the service config, albeit highly related. I do remember there was a point that failing for unknown filter keys was part of the design and I remember discussing alternatives, but I don't remember which alternative was chosen. For example, one option that I think was discussed is to ignore the _entry_ if there is an unknown filter, and go check other entries.

I'm fine with simply ignoring other fields (especially when considering proto), but at least a quick discussion would be good.

**Comment by dfawley** [comment]
Like C, Go's implementation currently ignores unknown fields in the list entries.

I can see it being problematic if an unknown filter key is ignored, because that key may have prevented the config from being selected.  However, it's equally problematic to skip the entry, for similar reasons, and it's obviously very bad to fail the entire list due to a single entry.  So, I'm OK with the C/Go behavior, with the expectation that service owners configuring this need to be aware of this behavior and the limitations of older client versions.

**Comment by AntonInglebert** [comment]
Hey, 
I am trying to get involved, could I try implementing this?
Instead of failing on a bad entry, keep going and warn ?

**Comment by YifeiZhuang** [comment]
@ejona86 This statement (might be the reason about java's behaviour?) is not accurate to me:
>  // If any unexpected field name is present in this object, the entire
    // config is considered invalid.

It is from [this section](https://github.com/grpc/proposal/blob/master/A2-service-configs-in-dns.md#canarying-changes)

**Comment by ejona86** [comment]
@markdroth, thoughts about @YifeiZhuang's most recent comment? Context since this is old: We noticed Java didn't allow any other keys in the filter matching for service config in DNS. Based on your and Doug's comments, C and Go allow other fields.

**Comment by ejona86** [comment]
I think we all agree the current language is just bad, because it limits you forever. The choices were to ignore unknown fields or to skip the choice. So Carl's PR makes a lot of sense to me, in multiple ways.  I think Yash's change makes sense for the `serviceConfig` field, but was a mistake for the unexpected field names.

Given the design and the history, I'd favor reverting the "unexpected field name" change in Yash's PR. If we went this route, we should probably reword it to avoid "invalid," so maybe "Any unexpected fields present in this object match no clients." Given C and Go's current behavior, it does add some weight to ignoring unknown fields, but it seems more like bugs or outdated implementations.

> We had basically this same concern when we discussed the xDS route matching selectors

For xDS there was no alternative. There's also a more limited timeline of support for xDS and client versions can be monitored by the control plane.

I do think ignoring the choice is much better than ignoring the fields. It adds a very good strategy for adding new fields. Ignoring fields doesn't really have a strategy for adding new fields other than "wait years for all clients to update, and then cross your fingers."

Java's current (mis)behavior doesn't really change the calculus, since the behavior prevents adding any new fields entirely... except that Java and C don't have DNS TXT enabled by default. So Go's current behavior is the only one that would be widely deployed.

**Comment by dfawley** [comment]
> I've added an item for the TLs to discuss on the 30th.

We still have not had time to discuss this issue, and may not for some time.  Given the lack of user complaints here, I think this is going to be shelved until it becomes a real concern.
